### ISSUE TITLE
Calls from a machine with two network adapers.

### ISSUE THREAD

**Issue by AverkinSergei** [siniypin/pjsip4net#50]
Hi!
I have some windowsforms application based on your libruary pjsip4net, and I try to call from a machine with two network adapters. My app send register request:
REGISTER sip:192.168.0.1:5060 SIP/2.0
Via: SIP/2.0/UDP 172.28.0.137:5060;rport;branch=z9hG4bKPj7bd33b361aea44968f29b70638040877
Max-Forwards: 70
From: <sip:2003@192.168.0.1>;tag=029b4a13a725407ab197a346736ff789
To: <sip:2003@192.168.0.1>
Call-ID: 4cc0824452014170962904d193f0f232
CSeq: 18509 REGISTER
Contact: <sip:2003@192.168.0.13:5060;transport=UDP>
Expires: 300
Content-Length:  0
So, via header contains IP-address 172.28.0.137 from first network adapter, but I need the 192.168.0.13 from second network adapter. I tried to disable the first adapter and my application starts to work properly, but I need both adapters.
Can I configure my app for work with the second adapter.
Reguards!

**Comment by siniypin** [comment]
Hello,

can you check the address your UDP transport is bound to? You could also share your config, it helps.

Cheers

**Comment by AverkinSergei** [comment]
This is my App.config:
<?xml version="1.0" encoding="utf-8" ?>
<configuration>
  <configSections>
    <section name="sipua" type="pjsip4net.Configuration.SipUserAgentSettingsSection, pjsip4net, Culture=neutral, PublicKeyToken=null" />
    <sectionGroup name="common">
      <section name="logging" type="Common.Logging.ConfigurationSectionHandler, Common.Logging" />
    </sectionGroup>
    <section name="log4net" type="log4net.Config.Log4NetConfigurationSectionHandler, log4net, Version=1.2.10.0, Culture=neutral, PublicKeyToken=1b44e1d426115821" />
  </configSections>
  <sipua srtp="Enabled" secureSignaling="0" logMessages="true" logLevel="5" traceAndDebug="true" autoAnswer="false" autoRecording="false" autoConference="false" maxCalls="5">
    
    <sipTransport type="udp" Port="5060" boundAddress="192.168.0.13" publicAddress="192.168.0.13"/>
    
    <media isVADEnabled="true" playbackDeviceId="7" captureDeviceId="5" />
  </sipua>
  <common>
    <logging>
      <factoryAdapter type="Common.Logging.Log4Net.Log4NetLoggerFactoryAdapter, Common.Logging.Log4Net1210">
        <arg key="configType" value="EXTERNAL" />
      </factoryAdapter>
    </logging>
  </common>
  <log4net>
    <appender name="ColoredConsoleAppender" type="log4net.Appender.ColoredConsoleAppender">
      <mapping>
        <level value="ERROR" />
        <foreColor value="White" />
        <backColor value="Red, HighIntensity" />
      </mapping>
      <mapping>
        <level value="INFO" />
        <backColor value="Green" />
      </mapping>
      <mapping>
        <level value="WARN" />
        <backColor value="Yellow" />
      </mapping>
      <layout type="log4net.Layout.PatternLayout">
        <conversionPattern value="%date [%thread] %-5level - %message%newline" />
      </layout>
    </appender>
    <appender name="FileAppender" type="log4net.Appender.FileAppender">
      <file value="siptrace.txt" />
      <appendToFile value="true" />
      <layout type="log4net.Layout.PatternLayout">
        <conversionPattern value="%date [%thread] %-5level - %message%newline" />
      </layout>
    </appender>

    <root>
      <level value="DEBUG" />
      <appender-ref ref="ColoredConsoleAppender" />
      <appender-ref ref="FileAppender" />
    </root>
  </log4net>
  <startup>
    <supportedRuntime version="v4.0" sku=".NETFramework,Version=v4.0" />
  </startup>
  <runtime>
    <assemblyBinding xmlns="urn:schemas-microsoft-com:asm.v1">
      <dependentAssembly>
        <assemblyIdentity name="System.Core" publicKeyToken="b77a5c561934e089" culture="neutral" />
        <bindingRedirect oldVersion="0.0.0.0-4.0.0.0" newVersion="4.0.0.0" />
      </dependentAssembly>
      <dependentAssembly>
        <assemblyIdentity name="Common.Logging" publicKeyToken="af08829b84f0328e" culture="neutral" />
        <bindingRedirect oldVersion="0.0.0.0-3.1.0.0" newVersion="3.1.0.0" />
      </dependentAssembly>
    </assemblyBinding>
  </runtime>
</configuration>

**Comment by AverkinSergei** [comment]
This is my App.config:
<?xml version="1.0" encoding="utf-8" ?>
\<configuration\>
  \<configSections\>
    \<section name="sipua" type="pjsip4net.Configuration.SipUserAgentSettingsSection, pjsip4net, Culture=neutral, PublicKeyToken=null" /\>
    \<sectionGroup name="common"\>
      \<section name="logging" type="Common.Logging.ConfigurationSectionHandler, Common.Logging" /\>
    \</sectionGroup>
    \<section name="log4net" type="log4net.Config.Log4NetConfigurationSectionHandler, log4net, Version=1.2.10.0, Culture=neutral, PublicKeyToken=1b44e1d426115821" /\>
  \</configSections\>
  \<sipua srtp="Enabled" secureSignaling="0" logMessages="true" logLevel="5" traceAndDebug="true" autoAnswer="false" autoRecording="false" autoConference="false" maxCalls="5"\>
    \<!--\<accounts\>
      \<account accountId="sip:2003@192.168.0.1:5060" registrarUri="sip:192.168.0.1:5060"
          userName="2003" password="1234" realm="asterisk" isDefault="true"
          publishPresence="true" /\>
    \</accounts\>--\>
    \<sipTransport type="udp" Port="5060" boundAddress="192.168.0.13" publicAddress="192.168.0.13"/\>
    \<!--\<networkSettings natInSDP="true" forceLooseRoute="false"\>
      \<stun address=""/\>
      \<turn enabled="true" server="192.168.0.1" type="udp" userName="2003" password="1234" realm="asterisk" /\>
      \<ice enabled="true" noRTCP="false" /\>
      \<proxies\>
        \<add address="sip:192.168.0.1:5060"/\>
      \</proxies\>
    \</networkSettings\>--\>
    \<media isVADEnabled="true" playbackDeviceId="7" captureDeviceId="5" /\>
  \</sipua\>
  \<common\>
    \<logging\>
      \<factoryAdapter type="Common.Logging.Log4Net.Log4NetLoggerFactoryAdapter, Common.Logging.Log4Net1210"\>
        \<arg key="configType" value="EXTERNAL" /\>
      \</factoryAdapter\>
    \</logging\>
  \</common\>
  \<startup\>
    \<supportedRuntime version="v4.0" sku=".NETFramework,Version=v4.0" /\>
  \</startup\>
  \<runtime\>
    \<assemblyBinding xmlns="urn:schemas-microsoft-com:asm.v1"\>
      \<dependentAssembly\>
        \<assemblyIdentity name="System.Core" publicKeyToken="b77a5c561934e089" culture="neutral" /\>
        \<bindingRedirect oldVersion="0.0.0.0-4.0.0.0" newVersion="4.0.0.0" /\>
      \</dependentAssembly\>
      \<dependentAssembly\>
        \<assemblyIdentity name="Common.Logging" publicKeyToken="af08829b84f0328e" culture="neutral" /\>
        \<bindingRedirect oldVersion="0.0.0.0-3.1.0.0" newVersion="3.1.0.0" /\>
      \</dependentAssembly\>
    \</assemblyBinding\>
  \</runtime\>
\</configuration\>

**Comment by siniypin** [comment]
So, having this config stanza `<sipTransport type="udp" Port="5060" boundAddress="192.168.0.13" publicAddress="192.168.0.13"/>` does not help, right?

**Comment by AverkinSergei** [comment]
Yes, right.

**Comment by siniypin** [comment]
Did you try to enable ICE negotiation?

**Comment by AverkinSergei** [comment]
How can I do it?

**Comment by AverkinSergei** [comment]
If you meen \<ice enabled="true" noRTCP="true" /\> It's do not work.

**Comment by AverkinSergei** [comment]
May be it's a bug?
I find in code that:
namespace pjsip4net.Configuration
{
    public class CfgFileConfigurationProvider : IConfigurationProvider
    {
        private SipUserAgentSettingsSection _section;

        #region IConfigurationProvider Members

        public void Configure(IConfigurationContext context)
        {
            try
            {
                var config = context.Config;
                var mediaConfig = context.MediaConfig;
                var loggingConfig = context.LoggingConfig;
                var cfg = ConfigurationManager.OpenExeConfiguration(ConfigurationUserLevel.None);
                _section = (SipUserAgentSettingsSection) cfg.Sections["sipua"];
                if (_section != null)
                {
                    context.RegisterTransport(
                        new Core.Utils.Tuple<TransportType, TransportConfig>(
                            (TransportType) Enum.Parse(typeof (TransportType), _section.Transport.TransportType, true),
                            new TransportConfig(){Port = (uint) _section.Transport.Port}));
**Here just the Port. no BoundAddress and no PublicAddress**

**Comment by AverkinSergei** [comment]
May be it's a bug?
I find in code that:
```
namespace pjsip4net.Configuration
{
    public class CfgFileConfigurationProvider : IConfigurationProvider
    {
        private SipUserAgentSettingsSection _section;

        #region IConfigurationProvider Members

        public void Configure(IConfigurationContext context)
        {
            try
            {
                var config = context.Config;
                var mediaConfig = context.MediaConfig;
                var loggingConfig = context.LoggingConfig;
                var cfg = ConfigurationManager.OpenExeConfiguration(ConfigurationUserLevel.None);
                _section = (SipUserAgentSettingsSection) cfg.Sections["sipua"];
                if (_section != null)
                {
                    context.RegisterTransport(
                        new Core.Utils.Tuple<TransportType, TransportConfig>(
                            (TransportType) Enum.Parse(typeof (TransportType), _section.Transport.TransportType, true),
                            new TransportConfig(){Port = (uint) _section.Transport.Port}));
```
**Here just the Port. no BoundAddress, no PublicAddress**

**Comment by AverkinSergei** [comment]
I want to explain ... The first interface 172.28.0.137 working with the Internet, the second 192.168.0.13 is working with a local intranet. Calls are made only on the local intranet.

**Comment by siniypin** [comment]
The code you mentioned (config stanzas translated to objects) may be a bug indeed. Could you give it a try with a fix? Should that fix the problem, please send me a pull request!

**Comment by AverkinSergei** [comment]
I think I found the solution. I just need to configure my UA in code, like this:
`ua.MediaManager.Provider.CreateMediaTransport(transportProvider.GetMediaTransportConfig());`
**transportProvider.GetMediaTransportConfig()**  is a custom method that defines a Port, BoundAddress and PublicAddress.

**Comment by siniypin** [comment]
That is correct, you can either [configure](https://github.com/siniypin/pjsip4net#configuration) your client with a [.config file](https://gist.github.com/siniypin/7860029)  or with a custom `IConfigurationProvider` [implementation](https://gist.github.com/siniypin/5951687) or with a [block of code](https://gist.github.com/siniypin/7860386).  
However, since you found a bug in a config provider, could you possibly fix it and send me a pull request? Everybody else would benefit from this fix as well.

**Comment by AverkinSergei** [comment]
Of course you're right, I'll try to fix this bug

**Comment by AverkinSergei** [comment]
Last revision is wrong! It don't work:
`namespace pjsip4net
{
    internal class DefaultSipUserAgent : Resource, ISipUserAgentInternal
    {
......
    localRegistry.RtpTransport =
                transportFactory
                .CreateTransport(TransportType.Udp,
                                 new TransportConfig()
                                 {
                                    BoundAddress = localRegistry.SipTransport.Config.BoundAddress,
                                    PublicAddress = localRegistry.SipTransport.Config.PublicAddress
                                 })
                .As<VoIPTransport>();
......`

It's my falt. It must be like this:
`localRegistry.RtpTransport =
                transportFactory
                .CreateTransport(TransportType.Udp,
                                 new TransportConfig())
                .As<VoIPTransport>();`

           ` using (localRegistry.RtpTransport.InitializationScope())
            {
                localRegistry.RtpTransport.Config.Port = 4000;
                localRegistry.RtpTransport.Config.BoundAddress = localRegistry.SipTransport.Config.BoundAddress;
                localRegistry.RtpTransport.Config.PublicAddress = localRegistry.SipTransport.Config.PublicAddress;
            }`

**Comment by siniypin** [comment]
Closed with #52
